subprojects/core/src/main/java/org/gradle/caching/internal/tasks/DefaultTaskOutputCachingBuildCacheKeyBuilder.java

```diff
@@ -107,23 +111,33 @@ private static void log(String name, Object value) {
     @Override
     public TaskOutputCachingBuildCacheKey build() {
         BuildCacheKeyInputs inputs = new BuildCacheKeyInputs(taskClass, classLoaderHash, actionClassLoaderHashes, actionTypes, inputHashes.build(), outputPropertyNames.build());
+        HashCode hash;
         if (classLoaderHash == null || actionClassLoaderHashes.contains(null)) {
-            return new DefaultTaskOutputCachingBuildCacheKey(null, inputs);
+            hash = null;
+        } else {
+            hash = hasher.hash();
         }
-        return new DefaultTaskOutputCachingBuildCacheKey(hasher.hash(), inputs);
+        return new DefaultTaskOutputCachingBuildCacheKey(taskPath, hash, inputs);
     }
 
     private static class DefaultTaskOutputCachingBuildCacheKey implements TaskOutputCachingBuildCacheKey {
 
+        private final Path taskPath;
```

Comment:
I find it strange that the `taskPath` is now part of the `BuildCacheKey`, given that we always emphasized that it is not part of the build cache key.